internal/envoy/v3/tracing.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	envoy_config_trace_v3 "github.com/envoyproxy/go-control-plane/envoy/config/trace/v3"
+	http "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	envoy_trace_v3 "github.com/envoyproxy/go-control-plane/envoy/type/tracing/v3"
+	envoy_type "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/projectcontour/contour/internal/dag"
+	"github.com/projectcontour/contour/internal/protobuf"
+	"github.com/projectcontour/contour/internal/timeout"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// TracingConfig returns a tracing config,
+// or nil if config is nil.
+func TracingConfig(tracing *EnvoyTracingConfig) *http.HttpConnectionManager_Tracing {
+	if tracing == nil {
+		return nil
+	}
+
+	var customTags []*envoy_trace_v3.CustomTag
+	for _, tag := range tracing.CustomTags {
+		if traceCustomTag := customTag(tag); traceCustomTag != nil {
+			customTags = append(customTags, traceCustomTag)
+		}
+	}
+
+	return &http.HttpConnectionManager_Tracing{
+		OverallSampling: &envoy_type.Percent{
+			Value: tracing.OverallSampling,
+		},
+		MaxPathTagLength: wrapperspb.UInt32(tracing.MaxPathTagLength),
```

Comment:
When I'm testing with a Contour config file, and I don't specify any `maxPathTagLength`, it's getting passed to Envoy as an explicit `0` which has the effect of always stripping the path from the `http.url` tag. Looks like this is becaue the field on Parameters is not a pointer, so the code can't distinguish between "not provided" and "explicitly set to 0". So, we should change that field to a pointer to be able to differentiate the two cases, and then _only_ pass a value to Envoy if the user explicitly specified a value. If they didn't specify a value, we don't need to explicitly default to 256, we can just not pass the field to Envoy and let its own default apply.

Reply:
Yes, this problem does occur, thanks for your review